crates/store/src/errors.rs

```diff
@@ -161,42 +172,61 @@ pub enum GenesisError {
+    ClosedChannel(RecvError),
     #[error("Concurrent write detected")]
     ConcurrentWrite,
+    #[error("Database doesn't have any block header data")]
+    DbBlockHeaderEmpty,
+    #[error("Database update task failed: {0}")]
+    DbUpdateTaskFailed(String),
+    #[error("Duplicated nullifiers {0:?}")]
+    DuplicatedNullifiers(Vec<Nullifier>),
+    #[error("Invalid output note type: {0:?}")]
+    InvalidOutputNoteType(Box<OutputNote>),
+    #[error("Invalid tx hash: expected {expected}, but got {actual}")]
+    InvalidTxHash { expected: RpoDigest, actual: RpoDigest },
+    #[error("Received invalid account tree root")]
+    NewBlockInvalidAccountRoot,
     #[error("New block number must be 1 greater than the current block number")]
     NewBlockInvalidBlockNum,
-    #[error("New block `prev_hash` must match the chain's tip")]
-    NewBlockInvalidPrevHash,
     #[error("New block chain root is not consistent with chain MMR")]
     NewBlockInvalidChainRoot,
-    #[error("Received invalid account tree root")]
-    NewBlockInvalidAccountRoot,
     #[error("Received invalid note root")]
     NewBlockInvalidNoteRoot,
     #[error("Received invalid nullifier root")]
     NewBlockInvalidNullifierRoot,
-    #[error("Duplicated nullifiers {0:?}")]
-    DuplicatedNullifiers(Vec<Nullifier>),
-    #[error("Block applying was broken because of closed channel on database side: {0}")]
-    BlockApplyingBrokenBecauseOfClosedChannel(RecvError),
-    #[error("Failed to create notes tree: {0}")]
-    FailedToCreateNoteTree(MerkleError),
-    #[error("Database doesn't have any block header data")]
-    DbBlockHeaderEmpty,
-    #[error("Failed to get MMR peaks for forest ({forest}): {error}")]
-    FailedToGetMmrPeaksForForest { forest: usize, error: MmrError },
-    #[error("Failed to update nullifier tree: {0}")]
-    FailedToUpdateNullifierTree(NullifierTreeError),
-    #[error("Invalid output note type: {0:?}")]
-    InvalidOutputNoteType(OutputNote),
-    #[error("Invalid tx hash: expected {expected}, but got {actual}")]
-    InvalidTxHash { expected: RpoDigest, actual: RpoDigest },
+    #[error("New block `prev_hash` must match the chain's tip")]
+    NewBlockInvalidPrevHash,
+}
+
+impl From<ApplyBlockError> for Status {
+    fn from(err: ApplyBlockError) -> Self {
+        match err {
+            ApplyBlockError::NewBlockInvalidBlockNum
+            | ApplyBlockError::NewBlockInvalidPrevHash
+            | ApplyBlockError::NewBlockInvalidChainRoot
+            | ApplyBlockError::NewBlockInvalidAccountRoot
+            | ApplyBlockError::NewBlockInvalidNoteRoot
+            | ApplyBlockError::NewBlockInvalidNullifierRoot
+            | ApplyBlockError::DuplicatedNullifiers(_)
+            | ApplyBlockError::InvalidOutputNoteType(_)
+            | ApplyBlockError::InvalidTxHash { .. } => Status::invalid_argument(err.to_string()),
+
+            _ => Status::internal(err.to_string()),
+        }
+    }
```

Comment:
This works, but I wonder if we should create a dedicated enum for these errors. For example, something like `InvalidBlockError`. Then, we could add `ApplyBlockError::InvalidBlockError` variant. I think the separation may be a bit cleaner.